src/session/session_op_tracker.c

```diff
@@ -0,0 +1,270 @@
+	 */
+	if (session->iface.connection == NULL)
+		return (0);
+
+	if (entry->api_boundary)
+		--session->api_call_depth;
+
+	WT_RET(__wt_epoch(session, &entry->end));
+	entry->done = 1;
+	/*
+	 * Set a timer in the parent if this operation is nested to facilitate
+	 * tracking self time.
+	 */
+	for (prev_entry = TAILQ_PREV(entry, __op_tracker, q);
+	    prev_entry != NULL;
+	    prev_entry = TAILQ_PREV(prev_entry, __op_tracker, q)) {
+		if (prev_entry->done)
+			continue;
+		memcpy(&prev_entry->last_start,
+		    &entry->end, sizeof(struct timespec));
+		break;
+	}
+
+	/* Reporting is done as we are returning from the API call. */
+	if (entry->api_boundary && session->api_call_depth == 0) {
+		/* Update the self timer since this is the end of the trace. */
+		entry->self_time_us +=
+		    WT_TIMEDIFF(entry->end, entry->last_start);
+		WT_RET(__wt_session_op_tracker_dump(
+		    session, session->op_trace_min));
+	}
+
+	WT_ASSERT(session, session->api_call_depth == 0 ||
+	    !TAILQ_EMPTY(&session->op_trackerq));
+	return (0);
+}
+
+/*
+ * __wt_session_op_tracker_clear --
+ *	Clear the slow op tracker. That is move all entries from the
+ *	active queue to the available queue.
+ */
+int
+__wt_session_op_tracker_clear(WT_SESSION_IMPL *session)
+{
+	WT_OP_TRACKER_ENTRY *entry;
+
+	WT_ASSERT(session, session->api_call_depth == 0);
+	while ((entry = TAILQ_FIRST(&session->op_trackerq)) != NULL) {
+		TAILQ_REMOVE(&session->op_trackerq, entry, q);
+		if (entry->msg != NULL)
+			__wt_scr_free(session, &entry->msg);
+		memset(entry, 0, sizeof(WT_OP_TRACKER_ENTRY));
+		TAILQ_INSERT_TAIL(&session->op_tracker_availq, entry, aq);
+	}
+
+	return (0);
+}
+
+/*
+ * __wt_session_op_tracker_destroy --
+ *	Free all memory associated with the slow operation tracker.
+ */
+int
+__wt_session_op_tracker_destroy(WT_SESSION_IMPL *session)
+{
+	WT_OP_TRACKER_ENTRY *entry;
+
+	TAILQ_FOREACH(entry, &session->op_trackerq, q) {
+		if (entry->msg != NULL)
+			__wt_scr_free(session, &entry->msg);
+		__wt_free(session, entry);
+	}
+	TAILQ_FOREACH(entry, &session->op_tracker_availq, aq) {
+		if (entry->msg != NULL)
+			__wt_scr_free(session, &entry->msg);
+		__wt_free(session, entry);
+	}
+
+	return (0);
+}
+
+/*
+ * __wt_session_op_tracker_setup --
+ *	Setup the slow operation tracking mechanism for this session.
+ */
+int
+__wt_session_op_tracker_setup(WT_SESSION_IMPL *session)
+{
+	TAILQ_INIT(&session->op_trackerq);
+	TAILQ_INIT(&session->op_tracker_availq);
+
+	return (0);
+}
+
+/*
+ * __wt_session_op_tracker_dump --
+ *	Write the tracking information for the last operation to the configured
+ *	message log.
+ */
+int
+__wt_session_op_tracker_dump(WT_SESSION_IMPL *session, uint64_t min_time)
+{
+	WT_DECL_RET;
+	WT_ITEM *buffer;
+	WT_OP_TRACKER_ENTRY *entry;
+	uint64_t cumulative_skipped_time, optime_us, self_time_us;
+
+	cumulative_skipped_time = 0;
+
+	if ((entry = TAILQ_FIRST(&session->op_trackerq)) == NULL)
+		return (WT_NOTFOUND);
```

Comment:
Do we really want to return WT_NOTFOUND here?  It seems like if there is nothing to do, we should just return 0, or if we really expect something to always be there, assert entry != NULL.  I don't think we want successful operations to suddenly start returning WT_NOTFOUND (like commit or rollback for example).